From patchwork Wed Aug 14 16:22:31 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Paolo Bonzini <pbonzini@redhat.com>
X-Patchwork-Id: 11094205
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 0FD2314F7
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:52 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id EFD0A28898
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:51 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id E41782889C; Wed, 14 Aug 2019 16:22:51 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 910A928897
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:51 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728325AbfHNQWl (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 14 Aug 2019 12:22:41 -0400
Received: from mail-wm1-f65.google.com ([209.85.128.65]:37220 "EHLO
        mail-wm1-f65.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727110AbfHNQWk (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 14 Aug 2019 12:22:40 -0400
Received: by mail-wm1-f65.google.com with SMTP id z23so4987180wmf.2;
        Wed, 14 Aug 2019 09:22:38 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=sender:from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=wZo+HAsrUj/Kw4UvtL/TEz9wPlQYNsu6PTm1383LfLY=;
        b=oP51eZQUXMmLRbDeCYOkf4P47J1oJKUutGh4AX/7V34VD3Xz/++eNtMFBLRBZw9tr9
         OISGpnQoBlkiAkQQxdmCeiL6+jY9KZFfu2Au+q5R6Jt0SygTlPP6rFXcTApEihP1KAt8
         fG5JxbCNA1bgv9YkQGOmwsYSYYooAz8jfXeQREv0P9tBMZfQBLH5C+LERJDaQUCMV5xc
         rAKzGqRbesEjbhCO1yf67HHOb7c2d0IB9pULgBEEUi+ze8NfIyufmKx6eDmln4DWt4Mq
         S8m4OGN0JD8a+TbcjcN4SdCRym4vbNvb569u+hxGcxR34MgRJ9BF6QU7jK1ceWgpJhGl
         NWPA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:sender:from:to:cc:subject:date:message-id
         :in-reply-to:references;
        bh=wZo+HAsrUj/Kw4UvtL/TEz9wPlQYNsu6PTm1383LfLY=;
        b=bjwYJe9ePLLtgHg/vX0W9z7tTmybsLM5im5byMc4rqRnta+jP7FaUzzkB3X96V0jmd
         XfT5CxGjfAC0c+BLfLkNufVFW4K1IarmghdVkqyTq58wfU8jgVkWtA5jbhHzOfIXiPlD
         BXcn5RO89+4hEHzcSuR+uHRAgY+/k1yn+wUPgRwH2z+bvnEgST0+Egdxbpn8Iv8Dds46
         KsDfYZSQQg1SZyXQ8MSsm/Dlenj3wPjE9PvWF7V6kLps9sp1UdJRUXhDH72Y7LdRSjbw
         ZFoF7MBC7EsXI/sM1Pzq3YeS6ve8qayJH9ubi8DWvzM2fqRJh91cgPZaCIdGB/y8eYeT
         W/9g==
X-Gm-Message-State: APjAAAU8iso1U15gyvAPF2wAGRiiXH6uoFRxX3kC3/9TWWboKKECFBNz
        qnIXm+883ibXfVb7YqLMH9+NEx2H
X-Google-Smtp-Source: 
 APXvYqxQ3LV9hhpjouPQzvw8y3xfnGyfDtK43vIvbpKIweDNLDlYskUTD/ziCK7fN9YRLFpvQz67zg==
X-Received: by 2002:a05:600c:24cb:: with SMTP id
 11mr904685wmu.94.1565799757758;
        Wed, 14 Aug 2019 09:22:37 -0700 (PDT)
Received: from 640k.localdomain.com ([93.56.166.5])
        by smtp.gmail.com with ESMTPSA id
 k124sm191620wmk.47.2019.08.14.09.22.36
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Wed, 14 Aug 2019 09:22:37 -0700 (PDT)
From: Paolo Bonzini <pbonzini@redhat.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: vkuznets@redhat.com
Subject: [PATCH 1/3] selftests: kvm: do not try running the VM in
 vmx_set_nested_state_test
Date: Wed, 14 Aug 2019 18:22:31 +0200
Message-Id: <1565799753-3006-2-git-send-email-pbonzini@redhat.com>
X-Mailer: git-send-email 1.8.3.1
In-Reply-To: <1565799753-3006-1-git-send-email-pbonzini@redhat.com>
References: <1565799753-3006-1-git-send-email-pbonzini@redhat.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

This test is only covering various edge cases of the
KVM_SET_NESTED_STATE ioctl.  Running the VM does not really
add anything.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---
 .../selftests/kvm/x86_64/vmx_set_nested_state_test.c      | 15 ---------------
 1 file changed, 15 deletions(-)

diff --git a/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c b/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c
index ed7218d166da..a99fc66dafeb 100644
--- a/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c
+++ b/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c
@@ -27,22 +27,13 @@
 
 void test_nested_state(struct kvm_vm *vm, struct kvm_nested_state *state)
 {
-	volatile struct kvm_run *run;
-
 	vcpu_nested_state_set(vm, VCPU_ID, state, false);
-	run = vcpu_state(vm, VCPU_ID);
-	vcpu_run(vm, VCPU_ID);
-	TEST_ASSERT(run->exit_reason == KVM_EXIT_SHUTDOWN,
-		"Got exit_reason other than KVM_EXIT_SHUTDOWN: %u (%s),\n",
-		run->exit_reason,
-		exit_reason_str(run->exit_reason));
 }
 
 void test_nested_state_expect_errno(struct kvm_vm *vm,
 				    struct kvm_nested_state *state,
 				    int expected_errno)
 {
-	volatile struct kvm_run *run;
 	int rv;
 
 	rv = vcpu_nested_state_set(vm, VCPU_ID, state, true);
@@ -50,12 +41,6 @@ void test_nested_state_expect_errno(struct kvm_vm *vm,
 		"Expected %s (%d) from vcpu_nested_state_set but got rv: %i errno: %s (%d)",
 		strerror(expected_errno), expected_errno, rv, strerror(errno),
 		errno);
-	run = vcpu_state(vm, VCPU_ID);
-	vcpu_run(vm, VCPU_ID);
-	TEST_ASSERT(run->exit_reason == KVM_EXIT_SHUTDOWN,
-		"Got exit_reason other than KVM_EXIT_SHUTDOWN: %u (%s),\n",
-		run->exit_reason,
-		exit_reason_str(run->exit_reason));
 }
 
 void test_nested_state_expect_einval(struct kvm_vm *vm,

From patchwork Wed Aug 14 16:22:32 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Paolo Bonzini <pbonzini@redhat.com>
X-Patchwork-Id: 11094207
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 2D65F14F7
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:57 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 1A26028896
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:57 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 0E0D2288A6; Wed, 14 Aug 2019 16:22:57 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 8C6CF28898
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:56 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728378AbfHNQWw (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 14 Aug 2019 12:22:52 -0400
Received: from mail-wr1-f66.google.com ([209.85.221.66]:33591 "EHLO
        mail-wr1-f66.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726265AbfHNQWl (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 14 Aug 2019 12:22:41 -0400
Received: by mail-wr1-f66.google.com with SMTP id u16so646229wrr.0;
        Wed, 14 Aug 2019 09:22:39 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=sender:from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=9IpXj+WLqrPfM7eBTih/QAL5m+fd30y7Bv1MfCGetRI=;
        b=KTV1AebBP6iE8e7HKKr6PbQEOuucGyw4XziuNXEtXV6KNoUxB96wCfPt5UjbM75/Eb
         hD21IfDpPH036DrvK0RAjAVX2V2cBLZ+mLfD9K2lmg1YXI3GTme8E6StDuAlhhMaQrBT
         XIHVOP/MbUS5juGxI94EGzlVjeJ/Xnu3rw2SXLjZ3jOHVWaqxalyISP0D+a0CvADw3mh
         vIerF9nbovPdkRHjH2CcyxgKkmZDiwGKi5RnXYdaCi69JPFQh6Cpkm4ilGvRVOBA9508
         wRHis7CpRcz9mqAtf1Y4hQkCMzeHGiEEDYydyVES2Zp2c8Ce/Ijg6flzvD0+fmQag+Ff
         3fZg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:sender:from:to:cc:subject:date:message-id
         :in-reply-to:references;
        bh=9IpXj+WLqrPfM7eBTih/QAL5m+fd30y7Bv1MfCGetRI=;
        b=cO+GVot3RCrwzJ6ZGOu7YB6Wnkz6EqgzQIprDCiviCenXkEfQwgb3fnYy0QFFCo7gY
         oYTC0d98YADXT4ytkFwx4XpWaGiM/P87Fx7b/p5Vz2aJDTncd4SZTg6HNWW1+KB5VCC+
         4v/thnk7aeoWGv461vQvThoEZVdrV5kPXDRwnV2Ps3hjhdaeXJtigattd0Og/O0laLF0
         1bdjn9i24XefUk1rxKGx2ZSdeHhg007G0zBacn7oOqm+E1w+CL2TOFsPKR/ugxL0I9sU
         P1FpllA21QLe8DxxjbMcdegoXaZIBcABBUZxNMj6PM5HYe/V7SgEjU9785RB4zKyVVj5
         EwJQ==
X-Gm-Message-State: APjAAAUS0yCn1my3Mf0RfRQfp9GrzCBql2L9N/VlcFdptcIo2r5b66i7
        pllGDyokHZdtMm8FbYeDinPopsuI
X-Google-Smtp-Source: 
 APXvYqx/OwnZ4kS12siXivbayZPya7LWMkiycRDASXEqVgLfDFKsEak0VHoShCNUFPeq+Ss1IgCxTQ==
X-Received: by 2002:a5d:628d:: with SMTP id k13mr558630wru.69.1565799758801;
        Wed, 14 Aug 2019 09:22:38 -0700 (PDT)
Received: from 640k.localdomain.com ([93.56.166.5])
        by smtp.gmail.com with ESMTPSA id
 k124sm191620wmk.47.2019.08.14.09.22.37
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Wed, 14 Aug 2019 09:22:38 -0700 (PDT)
From: Paolo Bonzini <pbonzini@redhat.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: vkuznets@redhat.com
Subject: [PATCH 2/3] selftests: kvm: provide common function to enable eVMCS
Date: Wed, 14 Aug 2019 18:22:32 +0200
Message-Id: <1565799753-3006-3-git-send-email-pbonzini@redhat.com>
X-Mailer: git-send-email 1.8.3.1
In-Reply-To: <1565799753-3006-1-git-send-email-pbonzini@redhat.com>
References: <1565799753-3006-1-git-send-email-pbonzini@redhat.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

There are two tests already enabling eVMCS and a third is coming.
Add a function that enables the capability and tests the result.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---
 tools/testing/selftests/kvm/include/evmcs.h       |  2 ++
 tools/testing/selftests/kvm/lib/x86_64/vmx.c      | 20 ++++++++++++++++++++
 tools/testing/selftests/kvm/x86_64/evmcs_test.c   | 15 ++-------------
 tools/testing/selftests/kvm/x86_64/hyperv_cpuid.c | 12 ++++--------
 4 files changed, 28 insertions(+), 21 deletions(-)

diff --git a/tools/testing/selftests/kvm/include/evmcs.h b/tools/testing/selftests/kvm/include/evmcs.h
index 4059014d93ea..4912d23844bc 100644
--- a/tools/testing/selftests/kvm/include/evmcs.h
+++ b/tools/testing/selftests/kvm/include/evmcs.h
@@ -220,6 +220,8 @@ struct hv_enlightened_vmcs {
 struct hv_enlightened_vmcs *current_evmcs;
 struct hv_vp_assist_page *current_vp_assist;
 
+int vcpu_enable_evmcs(struct kvm_vm *vm, int vcpu_id);
+
 static inline int enable_vp_assist(uint64_t vp_assist_pa, void *vp_assist)
 {
 	u64 val = (vp_assist_pa & HV_X64_MSR_VP_ASSIST_PAGE_ADDRESS_MASK) |
diff --git a/tools/testing/selftests/kvm/lib/x86_64/vmx.c b/tools/testing/selftests/kvm/lib/x86_64/vmx.c
index 204f847bd065..9cef0455b819 100644
--- a/tools/testing/selftests/kvm/lib/x86_64/vmx.c
+++ b/tools/testing/selftests/kvm/lib/x86_64/vmx.c
@@ -12,6 +12,26 @@
 
 bool enable_evmcs;
 
+int vcpu_enable_evmcs(struct kvm_vm *vm, int vcpu_id)
+{
+	uint16_t evmcs_ver;
+
+	struct kvm_enable_cap enable_evmcs_cap = {
+		.cap = KVM_CAP_HYPERV_ENLIGHTENED_VMCS,
+		 .args[0] = (unsigned long)&evmcs_ver
+	};
+
+	vcpu_ioctl(vm, vcpu_id, KVM_ENABLE_CAP, &enable_evmcs_cap);
+
+	/* KVM should return supported EVMCS version range */
+	TEST_ASSERT(((evmcs_ver >> 8) >= (evmcs_ver & 0xff)) &&
+		    (evmcs_ver & 0xff) > 0,
+		    "Incorrect EVMCS version range: %x:%x\n",
+		    evmcs_ver & 0xff, evmcs_ver >> 8);
+
+	return evmcs_ver;
+}
+
 /* Allocate memory regions for nested VMX tests.
  *
  * Input Args:
diff --git a/tools/testing/selftests/kvm/x86_64/evmcs_test.c b/tools/testing/selftests/kvm/x86_64/evmcs_test.c
index f95c08343b48..92915e6408e7 100644
--- a/tools/testing/selftests/kvm/x86_64/evmcs_test.c
+++ b/tools/testing/selftests/kvm/x86_64/evmcs_test.c
@@ -79,11 +79,6 @@ int main(int argc, char *argv[])
 	struct kvm_x86_state *state;
 	struct ucall uc;
 	int stage;
-	uint16_t evmcs_ver;
-	struct kvm_enable_cap enable_evmcs_cap = {
-		.cap = KVM_CAP_HYPERV_ENLIGHTENED_VMCS,
-		 .args[0] = (unsigned long)&evmcs_ver
-	};
 
 	/* Create VM */
 	vm = vm_create_default(VCPU_ID, 0, guest_code);
@@ -96,13 +91,7 @@ int main(int argc, char *argv[])
 		exit(KSFT_SKIP);
 	}
 
-	vcpu_ioctl(vm, VCPU_ID, KVM_ENABLE_CAP, &enable_evmcs_cap);
-
-	/* KVM should return supported EVMCS version range */
-	TEST_ASSERT(((evmcs_ver >> 8) >= (evmcs_ver & 0xff)) &&
-		    (evmcs_ver & 0xff) > 0,
-		    "Incorrect EVMCS version range: %x:%x\n",
-		    evmcs_ver & 0xff, evmcs_ver >> 8);
+	vcpu_enable_evmcs(vm, VCPU_ID);
 
 	run = vcpu_state(vm, VCPU_ID);
 
@@ -146,7 +135,7 @@ int main(int argc, char *argv[])
 		kvm_vm_restart(vm, O_RDWR);
 		vm_vcpu_add(vm, VCPU_ID);
 		vcpu_set_cpuid(vm, VCPU_ID, kvm_get_supported_cpuid());
-		vcpu_ioctl(vm, VCPU_ID, KVM_ENABLE_CAP, &enable_evmcs_cap);
+		vcpu_enable_evmcs(vm, VCPU_ID);
 		vcpu_load_state(vm, VCPU_ID, state);
 		run = vcpu_state(vm, VCPU_ID);
 		free(state);
diff --git a/tools/testing/selftests/kvm/x86_64/hyperv_cpuid.c b/tools/testing/selftests/kvm/x86_64/hyperv_cpuid.c
index f72b3043db0e..ee59831fbc98 100644
--- a/tools/testing/selftests/kvm/x86_64/hyperv_cpuid.c
+++ b/tools/testing/selftests/kvm/x86_64/hyperv_cpuid.c
@@ -18,6 +18,7 @@
 #include "test_util.h"
 #include "kvm_util.h"
 #include "processor.h"
+#include "vmx.h"
 
 #define VCPU_ID 0
 
@@ -106,12 +107,7 @@ int main(int argc, char *argv[])
 {
 	struct kvm_vm *vm;
 	int rv;
-	uint16_t evmcs_ver;
 	struct kvm_cpuid2 *hv_cpuid_entries;
-	struct kvm_enable_cap enable_evmcs_cap = {
-		.cap = KVM_CAP_HYPERV_ENLIGHTENED_VMCS,
-		 .args[0] = (unsigned long)&evmcs_ver
-	};
 
 	/* Tell stdout not to buffer its content */
 	setbuf(stdout, NULL);
@@ -136,14 +132,14 @@ int main(int argc, char *argv[])
 
 	free(hv_cpuid_entries);
 
-	rv = _vcpu_ioctl(vm, VCPU_ID, KVM_ENABLE_CAP, &enable_evmcs_cap);
-
-	if (rv) {
+	if (!kvm_check_cap(KVM_CAP_HYPERV_ENLIGHTENED_VMCS)) {
 		fprintf(stderr,
 			"Enlightened VMCS is unsupported, skip related test\n");
 		goto vm_free;
 	}
 
+	vcpu_enable_evmcs(vm, VCPU_ID);
+
 	hv_cpuid_entries = kvm_get_supported_hv_cpuid(vm);
 	if (!hv_cpuid_entries)
 		return 1;

From patchwork Wed Aug 14 16:22:33 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Paolo Bonzini <pbonzini@redhat.com>
X-Patchwork-Id: 11094203
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id A2F7113AC
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:44 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 8EBD528898
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:44 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 830F22889C; Wed, 14 Aug 2019 16:22:44 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 1F9FD288A7
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 14 Aug 2019 16:22:44 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728348AbfHNQWn (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 14 Aug 2019 12:22:43 -0400
Received: from mail-wr1-f65.google.com ([209.85.221.65]:43181 "EHLO
        mail-wr1-f65.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1728265AbfHNQWm (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 14 Aug 2019 12:22:42 -0400
Received: by mail-wr1-f65.google.com with SMTP id y8so5721356wrn.10;
        Wed, 14 Aug 2019 09:22:40 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=sender:from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=4b8CPfu268RbprWIAAnR9GOAkcKFH0eYu1KeRaQ2giI=;
        b=dWTk3SUIh/kNncCDReLtOQwHXQpQ0xfR1S76OsgRRNd/jeWkhr3yCFHns4FFw9Rd7j
         /+slB4R/F1jCtyAHpRJctNoXx4P0Y8hQfyTP/H5E/pWCfggJ/PKbk5/zm2ueZQIsiufC
         XqZASQAV6s/tDldopfn57UAzrYPYmvnZqihTfPcPB5PTGd3M21ghg3fd0ziHOwnWCf8e
         kAiTXDCrt0xRGyEM3MRIS7+ipcDAuXw0tMFGyT1UGd+HYBS9jihPhHpnqwKdR+wAOmpV
         ZMa8wDerJcU2eHOilVYnmadtLIFpRlRoJ3S+jfPSzwTGTuQ16epZE8nS9nAPCrVAHXGt
         QdFw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:sender:from:to:cc:subject:date:message-id
         :in-reply-to:references;
        bh=4b8CPfu268RbprWIAAnR9GOAkcKFH0eYu1KeRaQ2giI=;
        b=eBEqJZtycq+g+EyrPa342OfAg7UG8RLo0PQ2G1c6Q8doO/zgja9kjbByw/HFPWbV4W
         YOqScFve2d3JozbNf2zzvZjPovXGMXHxoXt7rLmckQevOFaNXCeGy04zdPgJ94cKnwxG
         1kgUB87Vhk/VKYKr0O43szUdrh+NZiGUJR5/vu5rgU5tu6Io8+P8dAXYPMENWd1d1qRz
         m6QAKolS7CS0ngCIZh1qOsgxmtHL+TBXxsatlOM0Z9PqRCzBxfGsWkLHg7OdU1MavQaM
         aKArQwRxLbz1ORCmqIHkWzbeWKTBRQ9U3ECCKT118xY2PH6PcKXp9CDtUx7jeSlc6aq4
         hZng==
X-Gm-Message-State: APjAAAVuJdgjpTyND94CkDP9TMPVrB3Rsbwh/tJ4Kfcilw83rtmtfrW6
        zN7twP07QzND4nfiupcusPo5/OGo
X-Google-Smtp-Source: 
 APXvYqwtFHsgIM0uzf0KOtLFWQuzo9lLmz0HOpJ6Ty6hJRjWM6hfiLWyJyAC/2FESOH4PdVwFh9NUw==
X-Received: by 2002:adf:ef07:: with SMTP id e7mr600507wro.242.1565799759725;
        Wed, 14 Aug 2019 09:22:39 -0700 (PDT)
Received: from 640k.localdomain.com ([93.56.166.5])
        by smtp.gmail.com with ESMTPSA id
 k124sm191620wmk.47.2019.08.14.09.22.38
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Wed, 14 Aug 2019 09:22:39 -0700 (PDT)
From: Paolo Bonzini <pbonzini@redhat.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: vkuznets@redhat.com
Subject: [PATCH 3/3] selftests: kvm: fix vmx_set_nested_state_test
Date: Wed, 14 Aug 2019 18:22:33 +0200
Message-Id: <1565799753-3006-4-git-send-email-pbonzini@redhat.com>
X-Mailer: git-send-email 1.8.3.1
In-Reply-To: <1565799753-3006-1-git-send-email-pbonzini@redhat.com>
References: <1565799753-3006-1-git-send-email-pbonzini@redhat.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

vmx_set_nested_state_test is trying to use the KVM_STATE_NESTED_EVMCS without
enabling enlightened VMCS first.  Correct the outcome of the test, and actually
test that it succeeds after the capability is enabled.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---
 .../selftests/kvm/x86_64/vmx_set_nested_state_test.c    | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c b/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c
index a99fc66dafeb..853e370e8a39 100644
--- a/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c
+++ b/tools/testing/selftests/kvm/x86_64/vmx_set_nested_state_test.c
@@ -25,6 +25,8 @@
 #define VMCS12_REVISION 0x11e57ed0
 #define VCPU_ID 5
 
+bool have_evmcs;
+
 void test_nested_state(struct kvm_vm *vm, struct kvm_nested_state *state)
 {
 	vcpu_nested_state_set(vm, VCPU_ID, state, false);
@@ -75,8 +77,9 @@ void set_default_vmx_state(struct kvm_nested_state *state, int size)
 {
 	memset(state, 0, size);
 	state->flags = KVM_STATE_NESTED_GUEST_MODE  |
-			KVM_STATE_NESTED_RUN_PENDING |
-			KVM_STATE_NESTED_EVMCS;
+			KVM_STATE_NESTED_RUN_PENDING;
+	if (have_evmcs)
+		state->flags |= KVM_STATE_NESTED_EVMCS;
 	state->format = 0;
 	state->size = size;
 	state->hdr.vmx.vmxon_pa = 0x1000;
@@ -126,13 +129,19 @@ void test_vmx_nested_state(struct kvm_vm *vm)
 	/*
 	 * Setting vmxon_pa == -1ull and vmcs_pa == -1ull exits early without
 	 * setting the nested state but flags other than eVMCS must be clear.
+	 * The eVMCS flag can be set if the enlightened VMCS capability has
+	 * been enabled.
 	 */
 	set_default_vmx_state(state, state_sz);
 	state->hdr.vmx.vmxon_pa = -1ull;
 	state->hdr.vmx.vmcs12_pa = -1ull;
 	test_nested_state_expect_einval(vm, state);
 
-	state->flags = KVM_STATE_NESTED_EVMCS;
+	state->flags &= KVM_STATE_NESTED_EVMCS;
+	if (have_evmcs) {
+		test_nested_state_expect_einval(vm, state);
+		vcpu_enable_evmcs(vm, VCPU_ID);
+	}
 	test_nested_state(vm, state);
 
 	/* It is invalid to have vmxon_pa == -1ull and SMM flags non-zero. */
@@ -217,6 +226,8 @@ int main(int argc, char *argv[])
 	struct kvm_nested_state state;
 	struct kvm_cpuid_entry2 *entry = kvm_get_supported_cpuid_entry(1);
 
+	have_evmcs = kvm_check_cap(KVM_CAP_HYPERV_ENLIGHTENED_VMCS);
+
 	if (!kvm_check_cap(KVM_CAP_NESTED_STATE)) {
 		printf("KVM_CAP_NESTED_STATE not available, skipping test\n");
 		exit(KSFT_SKIP);
